libtenzir/builtins/formats/kv.cpp

```diff
@@ -0,0 +1,197 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/arrow_table_slice.hpp>
+#include <tenzir/collect.hpp>
+#include <tenzir/operator_control_plane.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/series_builder.hpp>
+
+#include <arrow/api.h>
+#include <re2/re2.h>
+
+namespace tenzir::plugins::kv {
+
+namespace {
+
+constexpr auto docs = "https://docs.tenzir.com/docs/next/formats/kv";
+
+class splitter {
+public:
+  splitter() = default;
+
+  explicit splitter(located<std::string_view> pattern) {
+    auto regex = std::make_unique<re2::RE2>(pattern.inner,
+                                            re2::RE2::CannedOptions::Quiet);
+    if (not regex->ok()) {
+      diagnostic::error("could not parse regex: {}", regex->error())
+        .primary(pattern.source)
+        .note("regex must be supported by RE2")
+        .docs("https://github.com/google/re2/wiki/Syntax")
+        .throw_();
+    }
+    auto groups = regex->NumberOfCapturingGroups();
+    if (groups > 1) {
+      diagnostic::error("regex must have at most one capturing group")
+        .primary(pattern.source)
+        .docs(docs)
+        .throw_();
+    }
+    if (groups != 1) {
+      regex = std::make_unique<re2::RE2>(fmt::format("({})", pattern.inner),
+                                         re2::RE2::CannedOptions::Quiet);
+      if (not regex->ok()) {
+        diagnostic::error("internal error: regex could not be parsed "
+                          "after adding a capture group")
+          .primary(pattern.source)
+          .note(regex->error())
+          .throw_();
+      }
+    }
+    regex_ = std::move(regex);
+  }
+
+  auto split(std::string_view input) const
+    -> std::pair<std::string_view, std::string_view> {
+    TENZIR_ASSERT_CHEAP(regex_);
+    TENZIR_ASSERT_CHEAP(regex_->NumberOfCapturingGroups() == 1);
+    auto group = std::string_view{};
+    // Ideally, we would just pass `&group` to re2. However, this does not
+    // compile depending on the version. Same goes for `absl::string_view*`.
+    auto group_arg = re2::RE2::Arg{
+      &group,
+      [](const char* str, size_t n, void* dest) -> bool {
+        *static_cast<std::string_view*>(dest) = {str, n};
+        return true;
+      },
```

Comment:
`re2::StringPiece` will work regardless of re2+absl version (see `grok.cpp`).

Reply:
Thanks! :pray: